hibernate-core/src/main/java/org/hibernate/dialect/CockroachDialect.java

```diff
@@ -111,8 +115,42 @@ public CockroachDialect() {
 	}
 
 	public CockroachDialect(DialectResolutionInfo info) {
-		super(info);
-		driverKind = PostgreSQLDriverKind.determineKind( info );
+		this( createVersion( info ), PostgreSQLDriverKind.determineKind( info ) );
+		registerKeywords( info );
+	}
+
+	protected static DatabaseVersion createVersion( DialectResolutionInfo info ) {
+		DatabaseVersion databaseVersion = null;
+
+		if ( info.getDatabaseMetadata() != null ) {
+			try (java.sql.Statement s = info.getDatabaseMetadata().getConnection().createStatement() ) {
+				final ResultSet rs = s.executeQuery( "SELECT version()" );
+				if ( rs.next() ) {
+					databaseVersion = parseVersion( rs.getString( 1 ) );
+				}
+			}
+			catch (SQLException ex) {
+				// Ignore
+			}
+		}
+		if ( databaseVersion == null ) {
+			throw new HibernateException("Could not determine the CockroachDB version");
+		}
+		return databaseVersion;
+	}
+
+	public static DatabaseVersion parseVersion(String versionString ) {
+		Matcher m = Pattern.compile( "v[\\d]+(\\.[\\d]+)?(\\.[\\d]+)?" ).matcher( versionString );
+		if ( m.find() ) {
+			String[] versionParts = m.group().substring( 1 ).split( "\\." );
+			// if we got to this point, there is at least a major version, so no need to check [].length > 0
+			int majorVersion = Integer.parseInt( versionParts[0] );
+			int minorVersion = versionParts.length > 1 ? Integer.parseInt( versionParts[1] ) : 0;
+			int microVersion = versionParts.length > 2 ? Integer.parseInt( versionParts[2] ) : 0;
+
+			return new SimpleDatabaseVersion( majorVersion, minorVersion, microVersion);
+		}
+		return null;
```

Comment:
```suggestion
	protected static DatabaseVersion parseVersion(String versionString) {
		try {
			final String[] versionParts = versionString.split( "\\." );
			// if we got to this point, there is at least a major version, so no need to check [].length > 0
			int majorVersion = Integer.parseInt( versionParts[0].substring( 1 ) );
			int minorVersion = versionParts.length > 1 ? Integer.parseInt( versionParts[1] ) : 0;
			int microVersion = versionParts.length > 2 ? Integer.parseInt( versionParts[2] ) : 0;

			return new SimpleDatabaseVersion( majorVersion, minorVersion, microVersion );
		}
		catch ( NumberFormatException e ) {
			// ignore
		}
		return null;
```

Reply:
What happened to the regular expression? I assume that removing it was a lapsus (just in case I tested your change and as-is it definitely won't work).

Reply:
I'm ok with the the/catch (obviously), but I'm going to push back on removing the regex, sorru Christian ;)

Reply:
I don't like the `Pattern` stuff, but it works. It's either try-catch or the `Pattern` thing you did. It might not matter in this specific case, but I generally prefer avoiding to create unnecessary intermediate objects like a `Pattern` if it's not helpful for readability or anything.